lib/pymedphys/_dicom/connect/listen.py

```diff
@@ -0,0 +1,207 @@
+
+        # The application entity
+        self.ae = None
+
+    def start(self):
+
+        # Initialise the Application Entity
+        self.ae = AE(ae_title=self.ae_title)
+
+        # Add the supported presentation context
+        self.ae.add_supported_context(VerificationSOPClass)
+        for context in StoragePresentationContexts:
+            self.ae.add_supported_context(context.abstract_syntax)
+
+        handlers = [
+            (evt.EVT_C_MOVE, self.on_c_store),
+            (evt.EVT_C_STORE, self.on_c_store),
+            (evt.EVT_C_ECHO, self.on_c_echo),
+            (evt.EVT_ACCEPTED, self.on_association_accepted),
+            (evt.EVT_RELEASED, self.on_association_released),
+        ]
+
+        # Start listening for incoming association requests
+        self.ae.start_server((self.host, self.port), evt_handlers=handlers, block=False)
+
+    def stop(self):
+
+        if self.ae:
+            self.ae.shutdown()
+
+    def on_c_echo(self, _):
+        """Respond to a C-ECHO service request.
+        """
+        logging.debug("C-ECHO!")
+        return 0x0000
+
+    def on_association_accepted(self, _):
+        self.association_directory = None
+
+    def on_association_released(self, _):
+
+        if self.on_released_callback:
+            self.on_released_callback(self.association_directory)
+
+    def on_c_store(self, event):
+
+        dataset = event.dataset
+
+        # TODO: Should these be place in a config file?
+        mode_prefixes = {
+            "CT Image Storage": "CT",
+            "Enhanced CT Image Storage": "CTE",
+            "MR Image Storage": "MR",
+            "Enhanced MR Image Storage": "MRE",
+            "Positron Emission Tomography Image Storage": "PT",
+            "Enhanced PET Image Storage": "PTE",
+            "RT Image Storage": "RI",
+            "RT Dose Storage": "RD",
+            "RT Plan Storage": "RP",
+            "RT Structure Set Storage": "RS",
+            "Computed Radiography Image Storage": "CR",
+            "Ultrasound Image Storage": "US",
+            "Enhanced Ultrasound Image Storage": "USE",
+            "X-Ray Angiographic Image Storage": "XA",
+            "Enhanced XA Image Storage": "XAE",
+            "Nuclear Medicine Image Storage": "NM",
+            "Secondary Capture Image Storage": "SC",
+        }
+
+        try:
+            mode_prefix = mode_prefixes[dataset.SOPClassUID.name]
+        except KeyError:
+            mode_prefix = "UN"
+
+        suid = dataset.SeriesInstanceUID
+        series_dir = self.storage_directory / suid
+        series_dir.mkdir(exist_ok=True)
+
+        filename = "{0!s}.{1!s}".format(mode_prefix, dataset.SOPInstanceUID)
+        filepath = series_dir / filename
+
+        if filepath.exists:
+            # TODO Currently just logging this, since I'm unsure of the desired
+            # behaviour if the file has already been recieved.
+            logging.debug("DICOM file already exists, overwriting")
```

Comment:
Strongly recommend not overwriting with the same SOP Instance UID.
The implication is that someone could have made a decision based on the content of an SOP Instance and then have that change underneath them after the fact.
It's not an error if the data are the same.  It is an error or at least a problem if the data is not consistent.

Reply:
Thanks for the good advice. In that case I'll avoid overwriting it and log a warning that it was not written due to the file already existing.

Reply:
My question here is what would be an appropriate status code to return if we discarded the file due to the file already existing? I'm looking at the list here and it isn't immediately obvious to me: https://pynetdicom2.readthedocs.io/en/latest/statuses.html

Perhaps an A7XX code, out of resources?

Reply:
A thought is that the file on disk could be hashed, and the file just sent could be hashed. If they are the same as each other then everything can move on, no issue. However, if they are different, then an error code would be appropriate. Something like new DICOM file with same SOP instance UID provided as one already on disk, but files do not match.

Reply:
See http://dicom.nema.org/medical/dicom/current/output/chtml/part07/chapter_9.html#sect_9.1.1.1.9 
Perhaps: Invalid SOP Instance (0117H) - Indicates that the SOP Instance UID specified implied a violation of the UID construction rules.
In conjunction with Simon's comment.
But in general, receiving a duplicate is not in fact an error.
You just document your behaviour in your conformance statement.

You could also respond with:
Warning (Status value is Service Class specific) - This indicates that the peer DIMSE Service User was able to store the composite SOP Instance, but detected a probable error.  Where a Warning would have a status of: 0001 or Bxxx or 0107 or 0116

However, be aware that most implementations will take a status that isn't 0x0000 to be a failure and respond accordingly.

Reply:
Thanks for the great info. I've had a go at making this more graceful. I check the hash on the dicom converted to json string of the existing file and the incoming file. This would only hash the object attributes and not the meta data which I think is desired.

Hmmmm, as I write this I don't think this will work for pixel data, since that wouldn't be output to the json which is what is hashed. I can hash that separately, but am open to suggestion on how to do this better.

If the hashes match, the file is ignored but a 0x0000 status is returned. If it conflicts the file is also discarded but a warning is returned.

Reply:
If it conflicts might it be worth storing the file sent in a "debugging" folder or something of the like?

Reply:
> If it conflicts might it be worth storing the file sent in a "debugging" folder or something of the like?

That's a very good idea.
In general it is very useful to have a place to put stuff that failed for some reason for future analysis (or manual repair)

Reply:
Sounds good, I will implement this. From CTP I know it as quarantine :D

Reply:
> Hmmmm, as I write this I don't think this will work for pixel data, since that wouldn't be output to the json which is what is hashed. I can hash that separately, but am open to suggestion on how to do this better.

The to_json() function actually does output the pixel data, so this approach will work if the pixel data is conflicting after all.